scripts/backport.py

```diff
@@ -0,0 +1,371 @@
+                input: {
+                    pullRequestId: "$pr_id",
+                    mergeMethod: REBASE
+                }
+            ) {
+                clientMutationId
+            }
+        }"""
+    ).substitute(pr_id=pr_id)
+    run_query(query)
+
+
+# The token has to have the "access public repositories" permission, or else creating a PR returns 404.
+github = Github(os.environ.get("GITHUB_TOKEN"))
+
+# If we are running inside Github Action, will modify the main repo.
+source_remote = "origin"
+source_repo_name = os.environ.get("GITHUB_REPOSITORY")
+target_remote = source_remote
+target_repo_name = source_repo_name
+
+if not source_repo_name:
+    # We are running manually for debugging, probably want to modify a fork.
+    source_repo_name = "timescale/timescaledb"
+    target_repo_name = os.environ.get("BACKPORT_TARGET_REPO")
+    target_remote = os.environ.get("BACKPORT_TARGET_REMOTE")
+    if not target_repo_name or not target_remote:
+        print(
+            "Please specify the target repositories for debugging, using the "
+            "environment variables BACKPORT_TARGET_REPO (e.g. `timescale/timescaledb`) "
+            "and BACKPORT_TARGET_REMOTE (e.g. `origin`).",
+            file=sys.stderr,
+        )
+        sys.exit(1)
+
+print(
+    f"Will look at '{source_repo_name}' (git remote '{source_remote}') for bug fixes, "
+    f"and create the backport PRs in '{target_repo_name}' (git remote '{target_remote}')."
+)
+
+source_repo = github.get_repo(source_repo_name)
+target_repo = github.get_repo(target_repo_name)
+
+# Set git name and email corresponding to the token user.
+token_user = github.get_user()
+os.environ["GIT_COMMITTER_NAME"] = token_user.name
+os.environ[
+    "GIT_COMMITTER_EMAIL"
+] = f"{token_user.id}+{token_user.login}@users.noreply.github.com"
+print(
+    f"Will commit as {os.environ['GIT_COMMITTER_NAME']} <{os.environ['GIT_COMMITTER_EMAIL']}>"
+)
+
+# Fetch the sources
+subprocess.run(f"git fetch {source_remote}", shell=True, check=True)
+subprocess.run(f"git fetch {target_remote}", shell=True, check=True)
+
+# Find out what is the branch corresponding to the previous version compared to
+# main. We will backport to that branch.
+version_config = {
+    x[0].strip(): x[1].strip()
+    for x in [
+        line.split("=")
+        for line in subprocess.check_output(
+            f"git show {source_remote}/main:version.config",
+            shell=True,
+            text=True,
+        ).splitlines()
+        if line
+    ]
+}
+
+previous_version = version_config["update_from_version"]
+previous_version_parts = previous_version.split(".")
+previous_version_parts[-1] = "x"
+backport_target = ".".join(previous_version_parts)
+
+print(f"Will backport to {backport_target}")
+
+# Find out which commits are unique to main and target branch. Also build sets of
+# the titles of these commits. We will compare the titles to check whether a
+# commit was backported.
```

Comment:
Doesn't this risk missing some commits if they have similar titles?

Since we never modify the master, can't we use the SHA on the master to decide if we should backport the commit or not. We could add the original SHA either to the commit message using `interpret-trailer` or to a note using `git notes`. (I would prefer the former since that does not require special treatment to push and pull notes.)

Reply:
Similar -- no. Exactly the same titles -- yes, they will be missed. But how likely this is in practice? I think the bug fixes usually have very distinct titles. I like the titles because this doesn't require any parsing.

```
$ git log --oneline --pretty=format:%s origin/main | sort | uniq -c | sort -nr | head -100
     18 Prepare the repo for next development cycle
      5 Prepare the repo for the next development cycle
      5 Add missing gitignore entry
      4 Fix flaky pg_dump test
      4 Bump postgres versions used in CI
      3 Remove unused functions
      3 Fix typos in comments
      3 Fix memory overflow
      2 Update CHANGELOG
      2 Upadting docs
      2 Run pgindent on code
      2 Remove unused functions from utils.c
      2 Remove dead code
      2 Release 2.0.0
      2 Move enterprise features to community
      2 Make older_than / newer_than error message user friendly
      2 Fix various misspellings
      2 Fix issues discovered by coverity
      2 Fix flaky bgw_custom test
      2 Fix compression_ddl isolation test
      2 Fix CAgg on CAgg variable bucket size validation
      2 Fix assertion failure in cursor_fetcher_rewind
      2 cleanup
```

Reply:
Adding the SHA of the original commit as a trailer would not require any parsing either. It would also allow finding the original commit easy to make sure that no mistakes were done.

Reply:
Now I'm using `cherry-pick -x`, which adds `(cherry picked from commit c5e496a554e9f4d04578f39669108554c22c918d)` to the end of the message. I think we have to support this kind of trailer, because this is a git-standard thing that you can also use when cherry-picking manually.

So matching by SHA is going to require:
1) some minimal parsing (e.g. re.match);
2) human operators not forgetting to add `-x` when they cherry-pick manually -- this one I'm worried about the most;
3) some special handling when you actually do a different manual fix in the release branch. You'll have to pretend it's a cherry-pick not to confuse the automation.

Using titles solves all three problems, because it's easy and natural to use the same title for manual cherry-pick/separate fix, and instead of parsing I just do `git log --pretty=format:%s`.

Let's observe for a while how problematic this title matching is going to be?

Reply:
> Now I'm using `cherry-pick -x`, which adds `(cherry picked from commit c5e496a554e9f4d04578f39669108554c22c918d)` to the end of the message. I think we have to support this kind of trailer, because this is a git-standard thing that you can also use when cherry-picking manually.
> 
> So matching by SHA is going to require:
> 
>     1. some minimal parsing (e.g. re.match);

Yeah, this one is simpler.
 
>     2. human operators not forgetting to add `-x` when they cherry-pick manually -- this one I'm worried about the most;

Since human operators can be quite creative in what they are doing we're probably going to need some separate CI checks for release branches any way.
 
>     3. some special handling when you actually do a different manual fix in the release branch. You'll have to pretend it's a cherry-pick not to confuse the automation.

Not sure that is a big problem. If you do a manual fix, you would probably do something like this anyway:

```
$ git cherry-pick -x <whatever>
<hack hack hack>
$ git commit -a -m'My tweaks'
<test test test>
$ git rebase -i upstream/main
<merge it all, and keep the original message>
```

> Using titles solves all three problems, because it's easy and natural to use the same title for manual cherry-pick/separate fix, and instead of parsing I just do `git log --pretty=format:%s`.
> 
> Let's observe for a while how problematic this title matching is going to be?

Seems it's already decided. :)